### PR TITLE
(REVIEW ONLY) (SERVER-1365) Expired CA cert can't sign certs

### PR DESCRIPTION
**Pull request by kurtwall**
A CA with an expired CA certificate should not be able to sign certs.

At the moment, this test runs and fails, likely because I haven't written it correctly, not because an expired CA can sign certs. I manually confirmed it cannot.

### CODE REVIEW

**Comment by cprice404** [comment]
@kurtwall first issue here is that you're referencing a fixtures directory from dev-resources that doesn't appear to exist... maybe you just need to add your local files to the commit?

File: test/integration/puppetlabs/services/certificate_authority/expired_ca_test.clj [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+      (ssl-utils/cert->pem! ca-cert ssl-ca-cert)
+      (ssl-utils/cert->pem! ca-cert ssl-cert)
+      (ssl-utils/obj->pem! csr csr-file)
+
+      (bootstrap/with-puppetserver-running
+        app
+        {:ssl-ca-cert ssl-ca-cert
+         :ssl-cert    ssl-cert
+         :ssl-key     ca-private-key}
```
Comment (cprice404) [inline test/integration/puppetlabs/services/certificate_authority/expired_ca_test.clj]: I *think* that you need to use a nested map here, where these keys are all underneath a top-level section called `webserver`, like [this](https://github.com/kurtwall/puppetserver/blob/9bb819d032c33f4b38ad8ddc3ca0de6f183ce7b6/test/integration/puppetlabs/services/master/environment_classes_int_test.clj#L680-L682):

```clj
{:webserver {:ssl-ca-cert ssl-ca-cert
                     :ssl-cert ssl-cert
                     :ssl-key ca-private-key}
```

File: test/integration/puppetlabs/services/certificate_authority/expired_ca_test.clj [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+        (let [signed-cert (utils/sign-certificate
```
Comment (cprice404) [inline test/integration/puppetlabs/services/certificate_authority/expired_ca_test.clj]: Calling `utils/sign-certificate` here is just calling directly into the ssl-utils *library* code to sign the certificate.  You wouldn't need puppetserver to be running for this, and it probably skips over any checks that the server would do about whether or not it's OK to sign the cert.  What you want to do in here, while you have the server running, is to make an HTTP request to the server to ask it to sign the cert.  That'll probably look something like [what @erikPrime did in his test](https://github.com/kurtwall/puppetserver/blame/9bb819d032c33f4b38ad8ddc3ca0de6f183ce7b6/test/integration/puppetlabs/services/certificate_authority/certificate_authority_int_test.clj#L193-L204):

```clj
(let [response (http-client/put
                       (str "https://localhost:8140"
                            "puppet-ca/v1/certificate_status/test_cert_ca_true")
                       {:ssl-cert (str master-conf-dir "/ssl/ca/ca_crt.pem")
                        :ssl-key (str master-conf-dir "/ssl/ca/ca_key.pem")
                        :ssl-ca-cert (str master-conf-dir "/ssl/ca/ca_crt.pem")
                        :as :text
                        :body "{\"desired_state\": \"signed\"}"
                        :headers {"content-type" "application/json"}})]
        (is (= 409 (:status response)))
        ...)
```

**Comment by kurtwall** [comment]
@cprice404 Apparently my SSLEngine needs maintenance. At least it got me this far.
```
2016-07-01 14:05:53,660 WARN  [async-dispatch-2] [o.e.j.s.h.ContextHandler] Empty contextPath
javax.net.ssl.SSLHandshakeException: General SSLEngine problem
 at sun.security.ssl.Handshaker.checkThrown (Handshaker.java:1431)
    sun.security.ssl.SSLEngineImpl.checkTaskThrown (SSLEngineImpl.java:535)
    sun.security.ssl.SSLEngineImpl.writeAppRecord (SSLEngineImpl.java:1214)
    sun.security.ssl.SSLEngineImpl.wrap (SSLEngineImpl.java:1186)
    javax.net.ssl.SSLEngine.wrap (SSLEngine.java:469)
    org.apache.http.nio.reactor.ssl.SSLIOSession.doWrap (SSLIOSession.java:263)
    org.apache.http.nio.reactor.ssl.SSLIOSession.doHandshake (SSLIOSession.java:303)
    org.apache.http.nio.reactor.ssl.SSLIOSession.isAppInputReady (SSLIOSession.java:507)
    org.apache.http.impl.nio.reactor.AbstractIODispatch.inputReady (AbstractIODispatch.java:122)
    org.apache.http.impl.nio.reactor.BaseIOReactor.readable (BaseIOReactor.java:164)
    org.apache.http.impl.nio.reactor.AbstractIOReactor.processEvent (AbstractIOReactor.java:339)
    org.apache.http.impl.nio.reactor.AbstractIOReactor.processEvents (AbstractIOReactor.java:317)
    org.apache.http.impl.nio.reactor.AbstractIOReactor.execute (AbstractIOReactor.java:278)
    org.apache.http.impl.nio.reactor.BaseIOReactor.execute (BaseIOReactor.java:106)
    org.apache.http.impl.nio.reactor.AbstractMultiworkerIOReactor$Worker.run (AbstractMultiworkerIOReactor.java:590)
    java.lang.Thread.run (Thread.java:745)
Ran 1 test containing 1 assertion.
```

**Comment by cprice404** [comment]
@kurtwall that error usually means that your client and server certs don't match - like, they were issued by different CAs.

**Pull request closed without merging**